Automatic Scroll on Search added

4shutosh opened the pull request:
## :camera: Screenshots
<!-- Show us what you've changed, we love images. -->
<img width="443" alt="Screenshot 2023-03-18 at 4 43 24 PM" src="https://user-images.githubusercontent.com/53831987/226102211-3779ed63-7c41-4943-8a4c-9aba5f4e7489.png">


## :page_facing_up: Context
<!-- Why did you change something? Is there an [issue](https://github.com/ChuckerTeam/chucker/issues) to link here? Or an external link? -->
Scroll to Search content added, fixes #388 
![Untitled](https://user-images.githubusercontent.com/53831987/226104203-888e4d2d-9547-4bbc-8af1-e3f032a102b1.gif)


## :pencil: Changes
<!-- Which code did you change? How? -->
<!-- If your changes affect users somehow, be it a new feature, a bugfix or an API change please update the `Unreleased` section in the CHANGELOG.md file. -->
Scroll to search item in response screen added

## :paperclip: Related PR
<!-- PR that blocks this one, or the ones blocked by this PR -->

## :no_entry_sign: Breaking
<!-- Is there something breaking the API? Any class or method signature changed? -->
No

## :hammer_and_wrench: How to test
<!-- Is there a special case to test your changes? -->
Open Response Screen, ideally where the whole body covers more than the screen size, and search for something that is not visible in the current scroll state to see the change. 

After searching if found, the recycler view is automatically scrolled to the position, making it easy to test/evaluate the response. Earlier, just the text was being highlighted making it hard to find in a lengthy response.

## :stopwatch: Next steps
<!-- Do we have to plan something else after the merge? -->

4shutosh commented:
- No visual change from the previous search results, different color can be added for the navigating item, but is that really required? 
- For a single search result, the up and down arrow buttons are not visible and are enabled/disabled on the ends of the list of search results (Gray color for that)

Any suggestions for code improvements, behavior change are highly appreciated.

4shutosh commented:
Please take a look @vbuberen @cortinico

cortinico commented:
> Please take a look @vbuberen @cortinico

Thanks for sendign this over @4shutosh This week I'm quite busy with KotlinConf, I won't be able to review this before the end of the conference. I can check it just after 👍
I also know that @vbuberen was interested in implementing something like this in the past so maybe he has relevant insights on this PR

ArjanSM reviewed (commented):
Hi @4shutosh  👋 
Looks good mostly!
I just have made a few comments looking forward to everyone's opinion.

4shutosh commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -117,6 +126,32 @@ internal class TransactionPayloadFragment :
+            inputMethodManager.hideSoftInputFromWindow(view?.windowToken, 0)
+        }
+
+        val scrollToIndex = if (goNext) currentSearchScrollIndex + 1 else currentSearchScrollIndex - 1
```
Hey @ArjanSM  thanks for the suggestion, can you please help me understand what the above code does better, or is there any behavior change that I am unable to see?

cortinico replied:
I believe @ArjanSM code will loop through indexes, while yours won't.

4shutosh replied:
Minor change was required here, implemented to make it loop. Thanks, Arjan.

cortinico commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ This file follows [Keepachangelog](https://keepachangelog.com/) format.
 Please add your entries according to this format.
 
 ## Unreleased
+* added scroll to highlighted text search in response screen
```
Can you move this to Added?

Also:
```suggestion
* Added scroll to highlighted text search in response screen [#988]
```

and add a link a the bottom of the file

4shutosh replied:
Sure

cortinico commented on library/src/main/res/layout/chucker_fragment_transaction_payload.xml:
```diff
@@ -68,4 +68,36 @@
         app:constraint_referenced_ids="emptyPayloadImage,emptyPayloadTextView"
         tools:visibility="visible" />
 
+    <com.google.android.material.floatingactionbutton.FloatingActionButton
+        android:id="@+id/fab_to_next"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:backgroundTint="@color/chucker_fab_background_colour"
+        android:contentDescription="@string/chucker_scroll_buttons_for_search"
+        android:src="@drawable/chucker_ic_arrow_down"
+        android:visibility="gone"
+        app:fabSize="mini"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:tint="@color/chucker_color_on_primary"
+        tools:visibility="visible" />
+
+    <com.google.android.material.floatingactionbutton.FloatingActionButton
+        android:id="@+id/fab_to_previous"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="4dp"
+        android:backgroundTint="@color/chucker_fab_background_colour"
+        android:contentDescription="@string/chucker_scroll_buttons_for_search"
+        android:enabled="false"
+        android:rotation="180"
+        android:src="@drawable/chucker_ic_arrow_down"
+        android:visibility="gone"
+        app:fabSize="mini"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toStartOf="@id/fab_to_next"
+        app:tint="@color/chucker_color_on_primary"
+        tools:enabled="false"
+        tools:visibility="visible" />
+
```
> We could chose to display the search results info summary by extending the App Bar to also include the icons(chevrons) which could be used to traverse through the filtered search results.

This would be preferred IMHO. Specifically having the # of occurrences would be helpful (see my video in top comment also).

Also I would refrain from using `FloatingActionButton` as they're specific to MDC and I'm looking into having MDC being used as a `compileOnly` dependency. So using classes directly inside the XML file makes things a bit harder.

4shutosh replied:
Yeah, I agree.

Will do the changes to summarise the search results in the app bar, along with the navigation chevrons.

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -200,15 +235,56 @@ internal class TransactionPayloadFragment :
+            }
+        }
+
+        Handler(Looper.getMainLooper()).postDelayed(
```
Can we use coroutines here instead?

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -200,15 +235,56 @@ internal class TransactionPayloadFragment :
+            listOfScrollableIndex.isEmpty() -> {
+                // scroll to top
+                currentSearchScrollIndex = 0
+                payloadBinding.payloadRecyclerView.smoothScrollToPosition(0)
```
I'm unsure this is good from the UX point of view (i.e. the user is searching and you keep on scrolling up and down).

4shutosh replied:
I prefer the implemented way, quickly jumping to the position. However, I can think of two solutions here. 

We can increase the initial delay.

Or we can only scroll after the IME_ACTION_DONE or keyboard closed - along with the summary of search results in the app bar.

cortinico replied:
Sorry my comment was related to the re-scroll to top once `listOfScrollableIndex.isEmpty`

4shutosh replied:
Oh okay, yeah let's not scroll in that condition. Will do the change.

cortinico reviewed (changes requested):
Great work @4shutosh 

I've been playing a bit with this feature and I believe, as @ArjanSM suggested, we need a bit more of a visual clue on the # of occurrences.

For example, look at this flow where all the occurrences are next to each other:

https://user-images.githubusercontent.com/3001957/233837948-6e12646c-3a00-4bd3-9d64-859d02a17773.mp4

Having no visual clue makes it impossible to know if there are "more" occurrences later. Also, I think we might highlight differently the selected occurrence (like with a different color).

4shutosh commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt:
```diff
@@ -200,15 +235,56 @@ internal class TransactionPayloadFragment :
             payloadAdapter.resetHighlight()
         }
+
+        scrollableIndices.clear()
```
I think all the other cases are handelled where currentSearchScrollIndex is changed, will check again if this is required or not

cortinico commented on CHANGELOG.md:
```diff
@@ -15,6 +14,7 @@ Please add your entries according to this format.
 * Added ability to export transactions to a file programmatically, LOG or HAR.
 * GraphQL OperationName header to transaction title [#69], [#116]
 * Added support for Android 13 and notifications permission handling
+* Added scroll to highlighted text search in response screen [#988]eye
```
```suggestion
* Added scroll to highlighted text search in response screen [#988]
```

4shutosh replied:
My bad

cortinico commented:
@4shutosh can you also resolve the conflicts?

4shutosh commented:
yes, on it.

4shutosh commented:
Added green color for the search item in focus, the search summary is in app bar now, and the nav chevron icons allow to loop through all the search results. 
![ezgif com-resize (1)](https://user-images.githubusercontent.com/53831987/235311856-2f37b7e9-dc98-4327-83c3-8c4daea1bad4.gif)

cortinico commented:
This looks great! 👍
There are still a couple of issues that we need to fix before this can go in:

1. I was playing with it and I managed to let the app crash with a zero division.
```
Process: com.chuckerteam.chucker.sample, PID: 15747
                                                                                                    java.lang.ArithmeticException: divide by zero
                                                                                                    	at com.chuckerteam.chucker.internal.ui.transaction.TransactionPayloadFragment.onSearchScrollerButtonClick(TransactionPayloadFragment.kt:172)
                                                                                                    	at com.chuckerteam.chucker.internal.ui.transaction.TransactionPayloadFragment.onViewCreated$lambda$3(TransactionPayloadFragment.kt:155)
                                                                                                    	at com.chuckerteam.chucker.internal.ui.transaction.TransactionPayloadFragment.$r8$lambda$X8t2VmZq9zp3O9oB7hdlJ8Q2YEc(Unknown Source:0)
                                                                                                    	at com.chuckerteam.chucker.internal.ui.transaction.TransactionPayloadFragment$$ExternalSyntheticLambda2.onClick(Unknown Source:2)
                                                                                                    	at android.view.View.performClick(View.java:7455)
                                                                                                    	at android.view.View.performClickInternal(View.java:7432)
                                                                                                    	at android.view.View.access$3700(View.java:835)
                                                                                                    	at android.view.View$PerformClick.run(View.java:28810)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7842)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

This happens if you try to search for "active" in the `/post` request to `postman-echo`.

2. Multiple occurrences in the same line are not handled properly. Here they're shown as 3 but they should be 5. Highlightning is also off.

![Screenshot_1682795370](https://user-images.githubusercontent.com/3001957/235320366-79bebb95-e8ac-4a30-ad41-050624a4c949.png)

You can reproduce this on the `/xml` request

4shutosh commented:
Haven't tried to find a solution for point 2 yet, but not able to figure out 1 after quite a few tries. It happens only after we've done the search once in a transaction fragment, go back, and go to another request transaction screen. I believe something wrong with the instance of the fragment, or while making the copy of `listOfScrollableIndex`. 

Any help regarding the same will be highly appreciated.

4shutosh commented:
Both the points are fixed now @cortinico. 

Implemented substring level highlighting for the body lines of the response. 

My doubt regarding the misbehaving fragment instance was correct. `BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT` was interfering with lazy variables initialization inside the fragment. I've updated the deprecated `FragmentStatePagerAdapter` to `viewPager2` & `FragmentStateAdapter`. 

Please review, and let me know if any other changes are required. 
![gif](https://user-images.githubusercontent.com/53831987/235349005-9cb22c40-3575-4137-ba69-37b00e04b280.gif)

vbuberen commented:
Hey there. I am joining a bit late, but would like to leave some general feedback.

1. I believe that search results should appear under tabs instead of making tabs jump down. For example, see how my version of search looked like:
![search_improvement](https://user-images.githubusercontent.com/13467769/235355890-958422dd-9bbf-4d0c-8a77-0f086e758667.gif)

2. We specifically skipped an update to `ViewPager2` because it had issues. Since there were no more stable version updates I believe that the issue still exists unless some other dependency like Fragment had to do something with it.
See for a reference: https://github.com/ChuckerTeam/chucker/pull/159#issuecomment-727571823

cortinico commented:
> My doubt regarding the misbehaving fragment instance was correct. `BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT` was interfering with lazy variables initialization inside the fragment. I've updated the deprecated `FragmentStatePagerAdapter` to `viewPager2` & `FragmentStateAdapter`.

Please let's do this in a separate PR as @vbuberen mentioned

4shutosh commented:
Reverted to old viewpager implemented. Search summary moved to fragment, below the tab layout.
![Screenshot 2023-05-03 at 5 46 11 PM](https://user-images.githubusercontent.com/53831987/235912965-4050586f-aa88-411d-9794-f78f8b1ef277.png)

I suspect some misbehaving highlighting on the requests page, please check if possible. Will test it, and fix if found something. 

Please take a look @cortinico, let me know if any changes are required.

cortinico reviewed (approved):
> Please take a look @cortinico, let me know if any changes are required.

Looks great to me @4shutosh 
I've done some manual testing and all looks great to me!

trietbui85 commented:
Awesome @4shutosh @cortinico . 
Is it the last PR before release v4.0.0?
